packages/cardpay-reward-programs/cardpay_reward_programs/rules/staking.py

```diff
@@ -0,0 +1,107 @@
+    This rule rewards users with CARD.cpxd held in a depot in a monthly basis
+    """
+
+    def __init__(self, core_parameters, user_defined_parameters):
+        super(Staking, self).__init__(core_parameters, user_defined_parameters)
+
+    def set_user_defined_parameters(self, token, duration, interest_rate_monthly):
+        self.token = token
+        self.duration = duration
+        self.interest_rate_monthly = interest_rate_monthly
+
+    def sql(self, table_query, aux_table_query):
+        token_holder_table = table_query
+        safe_owner_table = aux_table_query
+        return f"""
+            with filtered_safes as (
+                SELECT safe, _block_number, type, max(owner) as owner
+                FROM {safe_owner_table} a
+                WHERE _block_number = (
+                    SELECT MAX(_block_number)
+                    FROM {safe_owner_table} b
+                    WHERE a.safe = b.safe
+                    AND _block_number::integer < $2::integer
+                )
+                group by safe, _block_number, type
+            ),
+            filtered_balances as (
+                select tht.safe, tht.balance_downscale_e9_uint64::int64 as balance_int64, tht._block_number, tht.token,
+                from {token_holder_table} as tht, filtered_safes as sot
+                where tht.safe = sot.safe
+                and sot.type = 'depot'
+            ),
+            balance_changes as (select
+                safe,
+                balance_int64 - lag(balance_int64, 1 ,0) over (partition by safe order by _block_number asc) as change,
+                $5::float+1 as interest_rate,
+                ($2::integer - _block_number::integer) / $4::float  as compounding_rate,
+                from filtered_balances
+                where _block_number::integer < $2::integer
+                and token = $3::text
+                and safe is not null
+                qualify
+                _block_number::integer >= $1::integer
+                ),
+            original_balances as (select safe,
+                last(balance_int64) as change,
+                $5::float+1 as interest_rate,
+                1 as compounding_rate,
+                from filtered_balances a
+                where _block_number::integer < $1::integer
+                and token = $3::text
+                and safe is not null
+                and _block_number = (
+                    SELECT MAX(_block_number)
+                    FROM filtered_balances b
+                    WHERE a.safe = b.safe
+                    AND _block_number::integer < $1::integer
+                )
+                group by safe
+                ),
+            all_data as (select * from original_balances union all select * from balance_changes)
+
+            select owner as payee, sum(change * ((interest_rate**compounding_rate) - 1)) as rewards
+            from all_data
+            left join filtered_safes using (safe)
+            group by payee
+        """
```

Comment:
This is real hard to grok. So im layout the data here. 

filtered_safes

    safe  _block_number   type   owner
0  safe1              0  depot  owner1
1  safe2              0  depot  owner2
2  safe3              0  depot  owner3

filtered_balances

    safe  balance_int64  _block_number   token
0  safe1           1000              0  card-0
1  safe1           1500             10  card-0
2  safe1           1750             20  card-0
3  safe1           1500             25  card-0
4  safe2           1000              0  card-0
5  safe2           1500             10  card-0
6  safe2           1200             15  card-0
7  safe2           1700             20  card-0
8  safe3           1000              0  card-0

balance_changes

    safe  change  interest_rate  compounding_rate
0  safe1    1000           1.06          1.000000
1  safe1     500           1.06          0.666667
2  safe1     250           1.06          0.333333
3  safe1    -250           1.06          0.166667
4  safe2    1000           1.06          1.000000
5  safe2     500           1.06          0.666667
6  safe2    -300           1.06          0.500000
7  safe2     500           1.06          0.333333
8  safe3    1000           1.06          1.000000

original_balances

Empty DataFrame
Columns: [safe, change, interest_rate, compounding_rate]
Index: []


all_data

    safe  change  interest_rate  compounding_rate
0  safe1    1000           1.06          1.000000
1  safe1     500           1.06          0.666667
2  safe1     250           1.06          0.333333
3  safe1    -250           1.06          0.166667
4  safe2    1000           1.06          1.000000
5  safe2     500           1.06          0.666667
6  safe2    -300           1.06          0.500000
7  safe2     500           1.06          0.333333
8  safe3    1000           1.06          1.000000


end_reward

    payee    rewards
0  owner3  59.999943
1  owner1  82.268583
2  owner2  80.742585

Reply:
@IanCal what is the compounding **period**??

Reply:
If there is a link to how you guys computed interest, it would be good so I can check by it.

Reply:
Might be a silly question. The rule has cycle length of 30 in the test. But here we see data from 0,10,15,20. Is this meant to be?

Reply:
> Might be a silly question. The rule has cycle length of 30 in the test. But here we see data from 0,10,15,20. Is this meant to be?

The users balance will change during the month, so we need to calculate based on that.

Reply:
> If there is a link to how you guys computed interest, it would be good so I can check by it.

Doing it by summing looking at the individual changes is in the RFC - other than that it's raising to a power of the number of periods. I can add code comments explaining what it's doing through the query

Reply:
> If there is a link to how you guys computed interest, it would be good so I can check by it.

Doing it by summing looking at the individual changes is in the RFC - other than that it's raising to a power of the number of periods. I can add code comments explaining what it's doing through the query.

I'd like to experiment with a tool that would let us split the query up and test the parts individually in some upcoming work.

Reply:
> > Might be a silly question. The rule has cycle length of 30 in the test. But here we see data from 0,10,15,20. Is this meant to be?
> 
> The users balance will change during the month, so we need to calculate based on that.

right this makes sense

Reply:
> > > Might be a silly question. The rule has cycle length of 30 in the test. But here we see data from 0,10,15,20. Is this meant to be?
> > 
> > 
> > The users balance will change during the month, so we need to calculate based on that.
> 
> right this makes sense

Think this just means that the compounding period is the payment cycle length